esp-config/src/generate.rs

```diff
@@ -0,0 +1,138 @@
+        *self = match self {
+            Value::Bool(_) => match s {
+                "false" => Value::Bool(false),
+                _ => Value::Bool(true), // TODO should existance of the key mean true?
+            },
+            Value::Number(_) => Value::Number(s.parse().map_err(|_| ParseError)?),
+            Value::String(_) => Value::String(String::from(s)),
+        };
+        Ok(())
+    }
+
+    fn as_string(&self) -> String {
+        match self {
+            Value::Bool(value) => String::from(if *value { "true" } else { "false" }),
+            Value::Number(value) => format!("{}", value),
+            Value::String(value) => value.clone(),
+        }
+    }
+}
+
+/// This function will parse any screaming snake case environment variables that
+/// match the given prefix. It will then attempt to parse the [`Value`]. Once
+/// the config has been parsed, this function will emit snake case cfg's
+/// _without_ the prefix which can be used in the dependant crate. After that,
+/// it will create a markdown table in the `OUT_DIR` under the name
+/// `{prefix}_config_table.md` where prefix has also been converted
+/// to snake case. This can be included in crate documentation to outline the
+/// available configuration options for the crate.
+///
+/// The tuple ordering for the`config` array is key, default, description.
+///
+/// Unknown keys with the supplied prefix will cause this function to panic.
+pub fn generate_config(prefix: &str, config: &[(&str, Value, &str)]) {
+    let mut prefix = screaming_snake_case(prefix);
+
+    let mut doc_table = String::from(TABLE_HEADER);
+    for (key, default, desc) in config {
+        let key = screaming_snake_case(key);
+        let default = default.as_string();
+        writeln!(doc_table, "|**{prefix}_{key}**|{desc}|{default}|").unwrap();
+    }
+
+    // only rebuild if build.rs changed. Otherwise Cargo will rebuild if any
+    // other file changed.
+    println!("cargo:rerun-if-changed=build.rs");
+
+    // Generate the template for the config
+    let mut configs = HashMap::new();
+    for (name, default, _) in config {
+        configs.insert(snake_case(name), default.clone());
+
+        // Rebuild if config envvar changed.
+        // Note this is currently buggy and requires a clean build - PR is pending https://github.com/rust-lang/cargo/pull/14058
+        println!("cargo:rerun-if-env-changed={name}");
+    }
+
+    // Try and capture input from the environment
+    for (var, value) in env::vars() {
+        if let Some(name) = var.strip_prefix(&format!("{prefix}_")) {
+            let name = snake_case(name);
+            // TODO should this be lowered to a warning for unknown configuration options?
+            // we could instead mark this as unknown and _not_ emit
+            // println!("cargo:rustc-check-cfg=cfg({})", name);
+            let Some(cfg) = configs.get_mut(&name) else {
+                panic!("Unknown env var {name}")
```

Comment:
Let's collect these in a vec and die at the end with a complete list.

Reply:
Any thoughts on the comment above it? We could just ignore it, but mark it as unknown meaning the unknown cfg lint would fire?

Reply:
I don't understand, why would the unknown cfg lint fire? The cfg in the hal will not have typos, and we can emit any sort of unknown cfgs - if the code doesn't use it, it won't trigger warnings I think?

Reply:
Ah yeah I had a misunderstanding of how that lint worked. It only triggers if the cfg is used in code, not if he cfg has been passed rustc (which would have made what I was suggesting work).

I'll collect the errors as you originally suggested.